src/gc.rs

```diff
@@ -0,0 +1,257 @@
+/// you pull in the overhead of a run-time garbage collector to manage and
+/// free `Gc` values behind the scenes.
+#[derive(PartialEq, Eq, Debug)]
+pub struct Gc<T: ?Sized> {
+    ptr: NonNull<GcBox<T>>,
+    _phantom: PhantomData<T>,
+}
+
+impl<T> Gc<T> {
+    /// Constructs a new `Gc<T>`.
+    pub fn new(v: T) -> Self {
+        Gc {
+            ptr: unsafe { NonNull::new_unchecked(GcBox::new(v)) },
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Constructs a new `Gc<MaybeUninit<T>>` which is capable of storing data
+    /// up-to the size permissible by `layout`.
+    ///
+    /// This can be useful if you want to store a value with a custom layout,
+    /// but have the collector treat the value as if it were T.
+    ///
+    /// `layout` must be at least as large as `T`, and have an alignment which
+    /// is the same, or bigger than `T`.
+    pub fn new_from_layout(layout: Layout) -> Option<Gc<MaybeUninit<T>>> {
+        let tl = Layout::new::<T>();
+        if layout.size() < tl.size() && layout.align() >= tl.align() {
+            return None;
+        }
+        Some(Gc::from_inner(GcBox::new_from_layout(layout)))
+    }
+}
+
+impl Gc<dyn Any> {
+    pub fn downcast<T: Any>(&self) -> Result<Gc<T>, Gc<dyn Any>> {
+        if (*self).is::<T>() {
+            let ptr = self.ptr.cast::<GcBox<T>>();
+            forget(self);
+            Ok(Gc::from_inner(ptr))
+        } else {
+            Err(Gc::from_inner(self.ptr))
+        }
+    }
+}
+
+impl<T: ?Sized> Gc<T> {
+    /// Get a raw pointer to the underlying value `T`.
+    pub fn into_raw(this: Self) -> *const T {
+        this.ptr.as_ptr() as *const T
+    }
+
+    pub fn ptr_eq(this: &Self, other: &Self) -> bool {
+        this.ptr.as_ptr() == other.ptr.as_ptr()
+    }
+
+    pub fn from_raw(raw: *const T) -> Gc<T> {
+        Gc {
+            ptr: unsafe { NonNull::new_unchecked(raw as *mut GcBox<T>) },
+            _phantom: PhantomData,
+        }
+    }
+
+    fn from_inner(ptr: NonNull<GcBox<T>>) -> Self {
+        Self {
+            ptr,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T> Gc<MaybeUninit<T>> {
+    /// As with `MaybeUninit::assume_init`, it is up to the caller to guarantee
+    /// that the inner value really is in an initialized state. Calling this
+    /// when the content is not yet fully initialized causes immediate undefined
+    /// behavior.
+    pub unsafe fn assume_init(self) -> Gc<T> {
+        let ptr = self.ptr.as_ptr() as *mut GcBox<MaybeUninit<T>>;
+        Gc::from_inner((&mut *ptr).assume_init())
+    }
+}
+
+impl<T: ?Sized + fmt::Display> fmt::Display for Gc<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&**self, f)
+    }
+}
+
+/// A `GcBox` is a 0-cost wrapper which allows a single `Drop` implementation
+/// while also permitting multiple, copyable `Gc` references. The `drop` method
```

Comment:
I'm not sure the second sentence ("the drop method") makes sense. Isn't the point of the interior `ManuallyDrop` to *guarantee* that `drop` isn't called on `T` except when we explicitly call `drop` ourselves?

Reply:
Argh, sorry, this comment is outdated. I'll remove the no longer relevant parts.